### PR TITLE
Gabi consumer rest

### PR DESCRIPTION
**Pull request by gabriel-tincu**
Functionality pertaining to the /consumer endpoints in kafka rest, plus some refactoring on the serialization / admin rest functionality

### CODE REVIEW

File: karapace/kafka_rest_apis.py [hunk 1]
```diff
@@ -592,67 +642,453 @@ def topic_details(self, content_type: str, *, topic: str):
...
+
+
+class ConsumerManager:
+    def __init__(self, config_path: str):
+        self.config = KarapaceBase.read_config(config_path)
+        self.hostname = f"http://{self.config['advertised_hostname']}:{self.config['port']}"
+        self.log = logging.getLogger("RestConsumerManager")
+        self.deserializer = SchemaRegistryDeserializer(config_path=config_path)
+        self.consumers = {}
+        self.consumer_formats = {}
```
Comment (vtainio) [inline karapace/kafka_rest_apis.py:690]: Not needed

File: karapace/kafka_rest_apis.py [hunk 2]
```diff
@@ -592,67 +642,453 @@ def topic_details(self, content_type: str, *, topic: str):
...
+        with self.consumer_locks[internal_name]:
+            # since on a subscribed topic, the actual topic list will get refreshed in the background,
+            # we can use the same logic here as the refresh code , and the actual poll code will ensure that
+            # the pattern is respected when it comes to fetch requests
+            consumer = self.consumers[internal_name].consumer
+            sub = consumer._subscription  # pylint: disable=W0212
+            if consumer.subscription() is None:
+                topics = []
+            elif not sub.subscribed_pattern:
+                topics = list(consumer.subscription())
```
Comment (vtainio) [inline karapace/kafka_rest_apis.py:915]: This should be used for all cases.

File: karapace/kafka_rest_apis.py [hunk 3]
```diff
@@ -592,67 +631,442 @@ def topic_details(self, content_type: str, *, topic: str):
...
+                not_found(message=f"Topic {topic} not found", content_type=content_type, sub_code=40401)
+            not_found(message=f"Partition {partition_id} not found", content_type=content_type, sub_code=40402)
+
+    def list_brokers(self, content_type: str):
+        metadata = self.cluster_metadata()
+        metadata.pop("topics")
+        self.r(metadata, content_type)
+
+
+class ConsumerManager:
```
[the remark this answers is not in the document]
Comment (gabriel-tincu) [reply karapace/kafka_rest_apis.py]: done

File: TODO.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
+### Divergences from confluent rex proxy:
```
[the remark this answers is not in the document]
Comment (gabriel-tincu) [reply TODO.md]: dropped for now, will handle with github issues

File: karapace/kafka_rest_apis.py [hunk 5]
```diff
@@ -1,42 +1,74 @@
...
 RECORD_CODES = [42201, 42202]
 FILTERED_TOPICS = {"__consumer_offsets"}
+KNOWN_FORMATS = {"json", "avro", "binary"}
+OFFSET_RESET_STRATEGIES = {"latest", "earliest"}
+
+TypedConsumer = namedtuple("TypedConsumer", ["consumer", "serialization_format", "config"])
+empty_response = partial(KarapaceBase.r, body={}, status=204, content_type="application/json")
+
+
+def internal_error(message, content_type):
```
[the remark this answers is not in the document]
Comment (gabriel-tincu) [reply karapace/kafka_rest_apis.py]: done

File: karapace/kafka_rest_apis.py [hunk 6]
```diff
@@ -592,67 +631,442 @@ def topic_details(self, content_type: str, *, topic: str):
...
+    def _assert(cond: bool, code: int, sub_code: int, message: str, content_type: str):
+        if not cond:
+            KarapaceBase.r(content_type=content_type, status=code, body={"message": message, "error_code": sub_code})
+
+    def _assert_consumer_exists(self, internal_name: str, content_type: str):
+        if internal_name not in self.consumers:
+            not_found(message=f"Consumer for {internal_name} not found", content_type=content_type, sub_code=40403)
+
+    @staticmethod
+    def _validate_offset_request_data(topic: str, partition_id: int, offset: int, cluster_metadata: dict):
```
[the remark this answers is not in the document]
Comment (gabriel-tincu) [reply karapace/kafka_rest_apis.py]: dropped, nice catch

File: karapace/kafka_rest_apis.py [hunk 7]
```diff
@@ -592,67 +631,442 @@ def topic_details(self, content_type: str, *, topic: str):
...
+                    bootstrap_servers=self.config["bootstrap_uri"],
+                    client_id=self.config["client_id"],
+                    security_protocol=self.config["security_protocol"],
+                    ssl_cafile=self.config["ssl_cafile"],
+                    ssl_certfile=self.config["ssl_certfile"],
+                    ssl_keyfile=self.config["ssl_keyfile"],
+                    group_id=group_name,
+                    fetch_min_bytes=request_data.get("fetch.min.bytes"),
+                    request_timeout_ms=request_data.get("consumer.request.timeout.ms"),
+                    enable_auto_commit="auto.commit.enable" not in request_data
```
[the remark this answers is not in the document]
Comment (gabriel-tincu) [reply karapace/kafka_rest_apis.py]: rewrote to a more readable form

File: karapace/kafka_rest_apis.py [hunk 8]
```diff
@@ -592,67 +631,442 @@ def topic_details(self, content_type: str, *, topic: str):
...
+                response["offsets"].append({
+                    "topic": tp.topic,
+                    "partition": tp.partition,
+                    "metadata": commit_info.metadata,
+                    "offset": commit_info.offset
+                })
+        KarapaceBase.r(body=response, content_type=content_type)
+
+    # SUBSCRIPTION
+
```
[the remark this answers is not in the document]
Comment (gabriel-tincu) [reply karapace/kafka_rest_apis.py]: fixed

**Comment by gabriel-tincu** [comment]
> Can you rebase this against master?

done

**Pull request merged**